sys/fw_slots/fw_slots.c

```diff
@@ -0,0 +1,477 @@
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ */
+
+/**
+ * @file
+ * @author      Mark Solters <msolters@gmail.com>
+ * @author      Francisco Acosta <francisco.acosta@inria.fr>
+ *
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include "fw_slots.h"
+#include "cpu_conf.h"
+#include "irq.h"
+#include "periph/flashpage.h"
+#include "hashes/sha256.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#define HASH_BUF             (1024)
+
+static uint8_t firmware_buffer[HASH_BUF];
+
+/**
+ * @brief       Read internal flash to a buffer at specific address.
+ *
+ * @param[in]   address - Address to be read.
+ * @param[in]   count - count in bytes.
+ *
+ * @param[out]  data_buffer - The buffer filled with the read information.
+ *
+ */
+static void int_flash_read(uint8_t *data_buffer, uint32_t address, uint32_t count)
+{
+    uint8_t *read_addres = (uint8_t*)address;
+    while (count--) {
+        *data_buffer++ = *read_addres++;
+    }
+}
+
+int fw_slots_validate_int_slot(uint8_t fw_slot)
+{
+    /*
+     * TODO
+     */
+
+    return 0;
+}
+
+uint32_t fw_slots_get_slot_address(uint8_t fw_slot)
+{
+    return get_slot_address(fw_slot);
+}
+
+uint32_t fw_slots_get_slot_page(uint8_t fw_slot)
+{
+    return get_slot_page(fw_slot);
+}
+
+void fw_slots_print_metadata(FW_metadata_t *metadata)
+{
+    printf("Firmware Size: %ld\n", metadata->size);
+    printf("Firmware Version: %#x\n", metadata->version);
+    printf("Firmware UUID: %#lx\n", metadata->uuid);
+    printf("Firmware HASH: ");
+    for (unsigned long i = 0; i < sizeof(metadata->hash); i++) {
+        printf("%02x ", metadata->hash[i]);
+    }
+    printf("\n");
+    printf("Firmware signed HASH: ");
+    for (unsigned long i = 0; i < sizeof(metadata->shash); i++) {
+        printf("%02x ", metadata->shash[i]);
+    }
+    printf("\n");
+}
+
+int fw_slots_get_int_metadata(uint8_t fw_slot_page, FW_metadata_t *fw_metadata)
+{
+    uint32_t fw_address;
+
+    fw_address = fw_slot_page * FLASHPAGE_SIZE + CPU_FLASH_BASE;
+
+    DEBUG("[fw_slots] Getting internal metadata on page %d at address %#lx\n",
+            fw_slot_page, fw_address);
+    int_flash_read((uint8_t*)fw_metadata, fw_address, sizeof(FW_metadata_t));
+
+    return 0;
+}
+
+int fw_slots_get_slot_metadata(uint8_t fw_slot, FW_metadata_t *fw_slot_metadata)
+{
+    /*
+     * TODO
+     */
+
+    return 0;
+}
+
+int fw_slots_get_int_slot_metadata(uint8_t fw_slot, FW_metadata_t *fw_slot_metadata)
+{
+    uint32_t page;
+
+    DEBUG("[fw_slots] Getting internal FW slot %d metadata\n", fw_slot);
+    if (fw_slot > MAX_FW_SLOTS || fw_slot == 0) {
```

Comment:
why `fw_slot == 0` is not valid? actually it would simplify the code quite a lot if the slots start at 0. So for 1 and 2 the variable should be 0 and 1. Or just rename so it's consistent with 0 and 1.